index.bs

```diff
@@ -8041,63 +8043,367 @@ ECMAScript Object values.
+    value of type |T|), perform the following steps:
+
+    1.  Let |value| be the result of [=converted to an ECMAScript value|converting=] |x| to an
+        ECMAScript value.
+    1.  Perform [=!=] [$Call$](|p|.\[[Resolve]], <emu-val>undefined</emu-val>, « |value| »).
+</div>
+
+<div algorithm>
+    To <dfn export for=Promise>reject</dfn> a <code>{{Promise}}&lt;<var ignore>T</var>&gt;</code>
+    |p| with reason |r| (an ECMAScript value), perform the following steps:
+
+    1.  Perform [=!=] [$Call$](|p|.\[[Reject]], <emu-val>undefined</emu-val>, « |r| »).
+</div>
+
+<div algorithm>
+
+    To <dfn id="dfn-perform-steps-once-promise-is-settled" export lt="upon settling">perform some steps once a promise is settled</dfn>,
+    given a <code>{{Promise}}&lt;|T|&gt;</code> |promise| and one or two sets of steps to perform,
+    covering when the promise is fulfilled, rejected, or both, perform the following steps:
+
+    1.  Let |onFulfilledSteps| be the following steps given argument |V|:
+        1.  Let |value| be the result of [=converted to an IDL value|converting=] |V| to an IDL
+            value of type |T|.
+        1.  If there are no steps that are required to be run if the promise was fulfilled, then
+            return <emu-val>undefined</emu-val>.
+        1.  Let |result| be the result of performing any steps that were required to be run if
+            the promise was fulfilled, given |value| if |T| is not {{void}}.
+        1.  Return |result|, [=converted to an ECMAScript value=].
+    1.  Let |onFulfilled| be [=!=] [$CreateBuiltinFunction$](|onFulfilledSteps|, « »):
+    1.  Let |onRejectedSteps| be the following steps given argument |R|:
         1.  Let |reason| be the result of [=converted to an IDL value|converting=]
             |R| to an IDL value of type {{any}}.
         1.  If there are no steps that are required to be run if the promise was rejected, then
             return <emu-val>undefined</emu-val>.
-        1.  Otherwise, return the result of performing any steps that were required to be run if the promise was rejected,
-            with |reason| as the rejection reason.
-    1.  Return [=!=] [$PerformPromiseThen$](|promise|, |onFulfilled|, |onRejected|).
+        1.  Let |result| be the result of performing any steps that were required to be run if the
+            promise was rejected, given |reason|.
+        1.  Return |result|, [=converted to an ECMAScript value=].
+    1.  Let |onRejected| be [=!=] [$CreateBuiltinFunction$](|onRejectedSteps|, « »):
+    1.  Let |constructor| be |promise|.\[[Promise]].\[[Realm]].\[[Intrinsics]].[[{{%Promise%}}]].
+    1.  Let |newCapability| be ? [$NewPromiseCapability$](|constructor|).
+    1.  Return [=!=] [$PerformPromiseThen$](|promise|.\[[Promise]], |onFulfilled|, |onRejected|, |newCapability|).
+
+    Note: If the provided steps return a value of type |T|, this algorithm returns a
+    <code>{{Promise}}&lt;<var ignore>T</var>&gt;</code>.
```

Comment:
Maybe we should use a different letter than T here, since it could be anything. And we should note that if they throw, it'll return a rejected promise? Should we also talk about what happens if they return promises?? Hmm, maybe just saying that it behaves like .then() would summarize better?